File: LeetCode/C/2-linked-list-两数相加.cpp
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */
#include "main.h"
// @lc code=start
class Solution {
  public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        auto p1 = l1, p2 = l2;
        ListNode res(-1), *p = &res;
        int carry = 0;
        while (p1 || p2 || carry != 0) {
            int a = p1 ? p1->val : 0;
            int b = p2 ? p2->val : 0;
            carry += a + b;
            p->next = new ListNode(carry % 10);
            carry /= 10;
            if (p1)
                p1 = p1->next;
            if (p2)
                p2 = p2->next;
            p = p->next;
        }
        return res.next;
    }
};
// @lc code=end

File: LeetCode/Go/2-linked-list-两数相加.go
/*
 * @lc app=leetcode.cn id=2 lang=golang
 *
 * [2] 两数相加
 */
package leetcode

// @lc code=start
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var carry = 0
	var res = &ListNode{}
	var p = res
	for l1 != nil || l2 != nil || carry != 0 {
		if l1 != nil {
			carry += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			carry += l2.Val
			l2 = l2.Next
		}
		p.Next = &ListNode{Val: carry % 10}
		carry /= 10
		p = p.Next
	}
	return res.Next
}

// @lc code=end
